scripts/merge-changelogs.ts

```diff
@@ -0,0 +1,191 @@
+import { writeFile, readFile } from 'fs/promises';
+import { GlobSync } from 'glob';
+
+const validMessageTypes = [
+  'Known Issue',
+  'Compatibility Note',
+  'New Functionality',
+  'Improvement',
+  'Fixed Issue',
+  'Updated Dependencies'
+] as const;
+
+interface Message {
+  packageNames: string[];
+  commit: string;
+  type: typeof validMessageTypes[number];
+  version: string;
+  message: string;
+}
+
+function getPackageName(changelog: string): string {
+  const packageNameRegex = /^# @sap-cloud-sdk\/(.+)$/m;
+  const title = changelog.match(packageNameRegex);
+
+  if (!changelog.startsWith('# @sap-cloud-sdk/') || !title || !title[1]) {
+    throw new Error('Changelog should have package name on first line');
+  }
+
+  return title[1];
+}
+
+function splitByVersion(changelog: string): any[] {
+  // Explaination: https://regex101.com/r/wCKs2A/2
+  const versionRegex = /## ([0-9]+\.[0-9]+\..+)([^]*?)(?=(\n## |$))/g;
+  const matches = changelog.matchAll(versionRegex);
+  const results: any[] = [];
+  for (const match of matches) {
+    results.push({
+      version: match[1],
+      content: match[2].trim()
+    });
+  }
+  return results;
+}
+
+function parseContent(
+  content: string,
+  version: string,
+  packageName: string
+): Message[] {
+  // I'm sorry... Explaination: https://regex101.com/r/9UhEwo/2
+  const contentRegex =
+    /- ((?<commit>.*):) (\[(?<type>.*?)\]) (?<message>[^]*?)(?=(\n- |\n### |$))|- Updated dependencies \[(?<depsCommit>.*)\](?<deps>[^]*?)\n*(?=(\n- |\n### |$))/g;
+  const results: Message[] = [];
+  for (const match of content.matchAll(contentRegex)) {
+    if (!match.groups?.commit && !match.groups?.depsCommit) {
+      throw new Error(
+        `No commit hash found for change set in ${packageName} for version ${version}!`
+      );
+    }
+    if (
+      !validMessageTypes.includes(match.groups?.type as any) &&
+      !match[0].includes('Updated dependencies')
+    ) {
+      throw new Error(
+        `Change set has an incorrect or missing type for change set in ${packageName} for version ${version}!`
+      );
+    }
+    const message = match.groups?.message?.trim() || match.groups?.deps?.trim();
+    if (message) {
+      results.push({
+        packageNames: [packageName],
+        version,
+        commit: match.groups?.commit || match.groups?.depsCommit,
+        type: (match.groups?.type as any) || 'Updated Dependencies',
+        message: match.groups?.message?.trim() || match.groups?.deps?.trim()
+      });
+    }
+  }
+  return results;
+}
+
+function parseChangelog(changelog: string): Message[] {
+  const packageName = getPackageName(changelog);
+  const versions = splitByVersion(changelog);
+  return versions
+    .map(({ version, content }) => parseContent(content, version, packageName))
+    .flat();
+}
+
+function writeMessagesOfType(
```

Comment:
[pp] I prefer the ?: operator for short things like return a==b ? 'foo' : 'bar' In this case the content is so long I would prefer a if() return and and empty as else/fallback